### PR TITLE
stm32H7  external NOR  ospi flash in  memorymapped mode

### PR DESCRIPTION
**Pull request by FRASTM**
This PR completes the https://github.com/zephyrproject-rtos/zephyr/pull/61082

to make the stm32h7 or stm32u5 with octo-NOR flash tested when the external NOR octoFlash is configured in MemoryMapped mode.

CONFIG_STM32_MEMMAP is enabling the external NOR flash MemoryMapped mode.

CONFIG_STM32_MEMMAP_EXT_FLASH_BASE_ADDRESS is defining the external nor flash base address

### CODE REVIEW

**Comment by FRASTM** [comment]
Tested with  stm32h7b3i_dk in MemoryMapped mode
requires mpu region definition
Writing few bytes in memorymapped mode is still possible

**Comment by GeorgeCGV** [comment]
@FRASTM, when used with a bootloader (_i.e. MCU-Boot_) shouldn't `init` be skipped?

If I understand it correctly:

* bootloader inits devices
* performs copy/move operations
* maps the memory
* jumps into it

The FW starts after the jump. There, the init sequence is going to be performed for all device and sys nodes.
In that case, the map shall be already enabled (_in the bootloader, as we "run" from it_) and not interrupted (_i.e. via a clock, xSPI re-init_).

Or did I miss something?

**Comment by FRASTM** [comment]
> Or did I miss something?

No, as you say I could face a pb of clocking. What I see is : 
The boot loader is correctly going to do_boot() in ..//bootloader/mcuboot/boot/zephyr/main.c
The reset vector is correctly going to the external NOR, 
the PC is correctly executing the application which is downloaded (with STM32CubeProgrammer) in the external NOR flash
but nothing on the output console. I am expecting "hello world from NOR flash"
```
*** Booting Zephyr OS build zephyr-v3.4.0-3121-g4df7ebf34d86 ***
I: Starting Direct-XIP bootloader
I: Image 0 Primary slot: Image not found
I: Secondary slot: version=0.0.0+0
I: Image 0 loaded from the secondary slot
I: Bootloader chainload address offset: 0x0
I: Jumping to the image slot
```

**Comment by GeorgeCGV** [comment]
> there is no xSPI node.

By xSPI I meant OSPI/QSPI.

What I am trying to say is that currently the flash is re-initialized after jump. Also, the #57337 kills the PLL1 clock during clock initialization (_after jump_). That affects XIP if the flash is clocked using it (_default_).

**Comment by FRASTM** [comment]
> 

In the sample application which is compiled linked and written for the external NOR flash, there is no oSPI flash  driver (no oSPI peripheral) This is just a _hello_word_ or a _basic/blinky_. About PLL and system clock I cannot tell. Once jumped to the NOR I cannot debug anymore after instruction
`((void (*)(void))vt->reset)();`
with `vt =  {msp = 0x24001100, reset = 0x9000173f}`

```
   0x9000173e:	movs	r0, #16
   0x90001740:	msr	BASEPRI, r0
   0x90001744:	ldr	r0, [pc, #36]	; (0x9000176c)
   0x90001746:	mov.w	r1, #2112	; 0x840
   0x9000174a:	adds	r0, r0, r1
   0x9000174c:	msr	PSP, r0
   0x90001750:	mrs	r0, CONTROL
   0x90001754:	movs	r1, #2
```

**Comment by ajhemphill91** [comment]
I was able to adapt samples/application_development/code_relocation_nocopy to work with the stm32h7b3i_dk based on this.

Created a stm32h7b3i_dk.conf in the boards dir there, containing:
```
CONFIG_FLASH=y
CONFIG_HAVE_CUSTOM_LINKER_SCRIPT=n
CONFIG_SPI=n
```

Our use-case was to load image data into the external flash and have it accessible to the user application. For this any external flash destined data is defined const inside of a file to be targeted by code relocation.

Then modify CMakeLists.txt to point the code relocation to the appropriate memory regions:
```
# Previously was EXTFLASH_TEXT, due to the linkerfile being used in this sample for nrf5340
zephyr_code_relocate(FILES src/ext_code.c LOCATION QSPI_TEXT NOCOPY)

# Also move .rodata section (const-defined image data)
zephyr_code_relocate(FILES src/ext_code.c LOCATION QSPI_RODATA NOCOPY)
```

Finally for any code/rodata relocated to external flash, you need a external loader configured appropriately to flash it. I was able to use the default MX25LM51245G_STM32H7B3I-DISCO.stldr provided with STM32CubeProgrammer. After the above changes, flashing the zephyr.hex with STM32CubeProgrammer properly uploads to the external flash.

**Comment by erwango** [comment]
> I was able to adapt samples/application_development/code_relocation_nocopy to work with the stm32h7b3i_dk based on this.

Quite interesting! Thanks @ajhemphill91 for sharing this work.
Would you be able to open an issue (or share a pull request) to discuss this more in details?

**Comment by FRASTM** [comment]
My test is to build the mcuboot in the internal flash (0x8000000) and a sample application for the NOR flash
`west build -p auto -b stm32h735g_disco samples/basic/blinky  --sysbuild -- -DCONFIG_XIP=y -DCONFIG_FLASH_BASE_ADDRESS=0x90000000 -DSB_CONFIG_BOOTLOADER_MCUBOOT=y`
 I have to set CONFIG_BOOT_DIRECT_XIP=y in the ../bootloader/mcuboot/boot/zephyr/boards/stm32h735g_disco.conf
I am using the STM32CubeProgrammer to flash the mcuboot @0x8000000 and the zephyr.signed.bin @ 0x90000000
```
*** Booting Zephyr OS build zephyr-v3.4.0-3599-g09ce3a6b8e3e ***
I: Starting Direct-XIP bootloader
I: Image 0 Primary slot: Image not found
I: Secondary slot: version=0.0.0+0
I: Image 0 loaded from the secondary slot
I: Bootloader chainload address offset: 0x0
I: Jumping to the image slot
```

**Comment by GeorgeCGV** [comment]
> My test is to build the mcuboot in the internal flash (0x8000000) and a sample application for the NOR flash
> west build -p auto -b stm32h735g_disco samples/basic/blinky  --sysbuild -- -DCONFIG_XIP=y -DCONFIG_FLASH_BASE_ADDRESS=0x90000000 -DSB_CONFIG_BOOTLOADER_MCUBOOT=y

I noticed that when the application is built with modified flash base address and QSPI device tree memory region is present, then MPU configures several regions for flash start address (`0x90000000`). One is named `FLASH` the other one is `QSPI` (_from DTC_). 

Simple overwrite of the QSPI memory attribute in the application overlay to be like RAM with execution allowed _(due to XIP being enabled_) like:
```
&quadspi_memory {                                                       
 	/delete-property/ zephyr,memory-attr;                                  
 	zephyr,memory-attr = <( DT_MEM_ARM(ATTR_MPU_RAM) )>;
};  
```
yields MPU exception. The one you mentioned:

As that doesn't solve MPU configuration with different attributes for the same start address (_didn't check the size_).

> Code is running in the external NOR flash but fails at 0x90000f74 in arm_core_mpu_enable

The QSPI memory region can be removed in the application overlay to avoid the exception:
`/delete-node/ &quadspi_memory;`

I didn't investigate what actual attributes are applied for the flash and how they differ from `ATTR_MPU_RAM`. Also, it might happens that FLASH MPU region has different size. Probably, configured MPU regions are not correct when the application is build with a different `FLASH_BASE_ADDRESS` and it matches the `QSPI` memory region.

**Comment by GeorgeCGV** [comment]
Had a bit of a look into:

> Code is running in the external NOR flash but fails at 0x90000f74 in arm_core_mpu_enable

From application perspective when it runs from mapped memory the FLASH starts at `CONFIG_FLASH_BASE_ADDRESS` that is `0x90000000`.

Flash region is defined in `zephyr\soc\arm\st_stm32\stm32h7\mpu_regions.c` as:
```
MPU_REGION_ENTRY("FLASH", CONFIG_FLASH_BASE_ADDRESS, REGION_FLASH_ATTR(REGION_FLASH_SIZE)),
```

However, the `REGION_FLASH_SIZE` doesn't have the same or bigger value than what is defined in DT for `quadspi_memory`.

The MPU initialization creates a hardcoded `FLASH` region (_from `mpu_regions.c`_), then creates the region from DT for the same starting address but of a way bigger size. That leads to a problem during the MPU-enable stage.

Deleting `quadspi_memory` node is not enough to fully fix the problem.

The `REGION_FLASH_SIZE` should be adjusted to match the main partition size or set to `quadspi_memory` size. That can be done using `CONFIG_FLASH_SIZE`. Keep in mind that the `FLASH` region uses `REGION_FLASH_ATTR` instead of `REGION_RAM_ATTR`.

**Comment by FRASTM** [comment]
thanks @GeorgeCGV, I will consider your comments

File: soc/arm/st_stm32/Kconfig [hunk 1]
```diff
@@ -24,6 +24,20 @@ config STM32_ENABLE_DEBUG_SLEEP_STOP
 	  effectivly destroys the use-case of `west attach`. Also
 	  SEGGER RTT and similar technologies need this.
 
+config EXT_FLASH_BASE_ADDRESS
```
Comment (erwango) [inline soc/arm/st_stm32/Kconfig:27]: We need to make this more specific to help users understanding why it's needed
```suggestion
config STM32_MEMMAP_EXT_FLASH_BASE_ADDRESS
```

File: boards/arm/b_u585i_iot02a/b_u585i_iot02a-common.dtsi [hunk 2]
```diff
@@ -131,11 +131,10 @@
 
 	status = "okay";
 
-	mx25lm51245: ospi-nor-flash@0 {
+	mx25lm51245: ospi-nor-flash@70000000 {
 		compatible = "st,stm32-ospi-nor";
-		reg = <0>;
+		reg = <0x70000000 DT_SIZE_M(64)>; /* 512 Mbits */
 		ospi-max-frequency = <DT_FREQ_M(50)>;
-		size = <DT_SIZE_M(512)>; /* 64 MBytes */
```
Comment (erwango) [inline boards/arm/b_u585i_iot02a/b_u585i_iot02a-common.dtsi:138]: If this is also valid in non memmap cases, let's make it a dedicated PR, extended to all similar descriptions in STM32 boards

File: boards/arm/b_u585i_iot02a/board.cmake [hunk 3]
```diff
@@ -12,7 +12,6 @@ if(CONFIG_BUILD_WITH_TFM)
   endif()
 endif()
 
-board_runner_args(stm32cubeprogrammer "--erase" "--port=swd" "--reset-mode=hw")
```
Comment (erwango) [inline boards/arm/b_u585i_iot02a/board.cmake:15]: Please check why it was there in the first place an explain why it is safe removing.

File: samples/boards/stm32/spi_flash/README.rst [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+Overview
+********
+
+This sample demonstrates using the flash API on a SPI NOR serial flash
+memory device.  While trivial it is an example of direct access to the
+external NOR quad or octo flash in MemoryMapped mode.
+The Nor quad/octo Flash is configured in MemoryMapped mode at the device init.
+Then read/arite access is done with memcopy.
+The sample is writing few bytes and read them back at offset 0
+Then read bytes at offset=SPI_FLASH_SECTOR_SIZE
```
Comment (erwango) [inline samples/boards/stm32/spi_flash/README.rst:15]: This kind of application is useful for our own needs, but not much for users.
Do we need to have it in tree ?

An MEMMAP XIP  basic hello world sample would be great though

**Review by erwango — CHANGES_REQUESTED** [review]

File: samples/boards/stm32/spi_flash/README.rst (hunk 4 above)
Comment (FRASTM) [reply samples/boards/stm32/spi_flash/README.rst:15]: A "MEMMAP xip application" is more complex to handle with --sysbuild building and linking, with memory partitioning and flashing to internal and external memories.
That sample is light : read few bytes memory mapped

File: boards/arm/b_u585i_iot02a/b_u585i_iot02a-common.dtsi (hunk 2 above)
Comment (FRASTM) [reply boards/arm/b_u585i_iot02a/b_u585i_iot02a-common.dtsi:138]: During the MemoryMapped mode, the driver controls the external flash base address and size.
Both values are given by this <reg> property (better solution than using a definition in the board Kconfig).
The <reg> property exists. Here it is filled with actual values of the external flash when the external flash is accessed in memorymapped mode.
When the external flash is accessed in indirect mode (non-MemoryMapped), the base address is useless and the size is given by the <size> property.

**Comment by FRASTM** [comment]
new testcase to validate the access to the external octoFlash in MemoryMapped mode
The CONFIG_STM32_MEMMAP is set to configure the exteranl octo-flash in MemoryMapped mode.
The CONFIG_STM32_MEMMAP_EXT_FLASH_BASE_ADDRESS  is defined for each bord

**Comment by FRASTM** [comment]
$ west build -p auto -b b_u585i_iot02a tests/drivers/flash/stm32/memmap

```
ospi-nor-flash@0 SPI flash testing in MemoryMapped
=================================
START - test_stm32_memmap_multi_sector
 PASS - test_stm32_memmap_multi_sector in 0.001 seconds
===================================================================
START - test_stm32_memmap_single_sector
 PASS - test_stm32_memmap_single_sector in 0.001 seconds
===================================================================
TESTSUITE memmap succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [memmap]: pass = 2, fail = 0, skip = 0, total = 2 duration = 0.002 seconds
 - PASS - [memmap.test_stm32_memmap_multi_sector] duration = 0.001 seconds
 - PASS - [memmap.test_stm32_memmap_single_sector] duration = 0.001 seconds                                                                                  
                                                                                                                                                             
------ TESTSUITE SUMMARY END ------                                                                                                                          
                                                                                                                                                             
===================================================================                                                                                          
PROJECT EXECUTION SUCCESSFUL
```

**Comment by FRASTM** [comment]
Example to build and link an application for the external NOR and execute it in XiP.
$ west build -p always -b stm32h7b3i_dk samples/boards/stm32/hello_world_xip/ --sysbuild -- -DCONFIG_FLASH_BASE_ADDRESS=0x90000000

**Comment by FRASTM** [comment]
It is also consist in : 
`west build -p always -b stm32h7b3i_dk samples/application_development/sysbuild/with_mcuboot --sysbuild -- -DCONFIG_FLASH_BASE_ADDRESS=0x90000000`
and flash 2 different binary  files to the 0x8000000 and 0x90000000, but it fails
```
*** Booting Zephyr OS build zephyr-v3.5.0-656-g2437cfca5cd6 ***
W: Failed reading sectors; BOOT_MAX_IMG_SECTORS=128 - too small?
W: Cannot upgrade: slots are not compatible
E: Unable to find bootable image
```

File: drivers/flash/flash_stm32_ospi.c [hunk 5]
```diff
@@ -949,12 +958,135 @@ static int stm32_ospi_mem_reset(const struct device *dev)
...
+					: HAL_OSPI_INSTRUCTION_16_BITS)
+				: HAL_OSPI_INSTRUCTION_16_BITS;
+	s_command.Instruction = (dev_cfg->data_rate == OSPI_STR_TRANSFER)
+				? ((dev_cfg->data_mode == OSPI_SPI_MODE)
+					? ((stm32_ospi_hal_address_size(dev) ==
+					HAL_OSPI_ADDRESS_24_BITS)
+						? SPI_NOR_CMD_READ_FAST
+						: SPI_NOR_CMD_READ_FAST_4B)
+					: SPI_NOR_OCMD_RD)
+				: SPI_NOR_OCMD_DTR_RD;
```
Comment (GeorgeCGV) [inline drivers/flash/flash_stm32_ospi.c:1007]: Why not using `read_opcode` and base other configuration on available `read_mode`, `data_mode`, `data_rate`?

**Comment by FRASTM** [comment]
rebasing on d7873e25fdd88c8740ef7ba9ef1ca5c3ef1bce5c
there is a pb with MPU on the stm32h7 : when the memmap mode of the external octo-flash is enabled, the read operation performed by a memcpy  results in a z_arm_usage_fault.
However reading the content of the flash @0x90000000  is correct
`0x90000000:	0x96f3b83d	0x00000000	0x00000400	0x00005808`

**Comment by FRASTM** [comment]
*** Booting Zephyr OS build zephyr-v3.5.0-3960-gf6dfcb10c7df ***
I: Starting Direct-XIP bootloader
I: Image 0 Primary slot: Image not found
I: Secondary slot: version=0.0.0+0
I: Image 0 loaded from the secondary slot
I: Bootloader chainload address offset: 0x0
I: Jumping to the image slot
I: NOR init'd in MemMapped mode

I: littlefs partition at /lfs1
I: LittleFS version 2.5, disk version 2.0
I: FS at flash-controller@52002000:0xe0000 is 8 0x2000-byte blocks with 512 cyce
I: sizes: rd 256 ; pr 256 ; ca 4096 ; la 256
I: Automount /lfs1 succeeded
*** Booting Zephyr OS build zephyr-v3.5.0-3960-gf6dfcb10c7df ***
Sample program to r/w files on littlefs
Area 0 at 0xe0000 on flash-controller@52002000 for 65536 bytes
/lfs1 automounted
/lfs1: bsize = 256 ; frsize = 8192 ; blocks = 8 ; bfree = 6

Listing dir /lfs1 ...
[FILE] boot_count (size = 1)
[FILE] pattern.bin (size = 547)
/lfs1/boot_count read count:11 (bytes: 1)
/lfs1/boot_count write new boot count 12: [wr:1]
------ FILE: /lfs1/pattern.bin ------
...
I: /lfs1 unmounted
/lfs1 unmount: 0

File: drivers/flash/flash_stm32_ospi.c (hunk 5 above)
Comment (FRASTM) [reply drivers/flash/flash_stm32_ospi.c:1002]: > Why not using `read_opcode` and base other configuration on available `read_mode`, `data_mode`, `data_rate`?

do you suggest : 
```
	s_command.Instruction = (dev_cfg->data_rate == OSPI_STR_TRANSFER)
				? ((dev_cfg->data_mode == OSPI_SPI_MODE)
					? ((stm32_ospi_hal_address_size(dev) ==
					HAL_OSPI_ADDRESS_24_BITS)
						? SPI_NOR_CMD_READ_FAST
						: SPI_NOR_CMD_READ_FAST_4B)
					: dev_data->read_opcode)
				: SPI_NOR_OCMD_DTR_RD;
```

File: soc/arm/st_stm32/Kconfig (hunk 1 above)
Comment (FRASTM) [reply soc/arm/st_stm32/Kconfig]: no more used, --> take the `DT_REG_ADDR(DT_INST(0, st_stm32_ospi_nor))`

File: drivers/flash/flash_stm32_ospi.c [hunk 6]
```diff
@@ -949,12 +958,135 @@ static int stm32_ospi_mem_reset(const struct device *dev)
...
+
+	/* Initialize the program command */
+	s_command.OperationType = HAL_OSPI_OPTYPE_WRITE_CFG;
+	if (dev_cfg->data_rate == OSPI_STR_TRANSFER) {
+		s_command.Instruction = (dev_cfg->data_mode == OSPI_SPI_MODE)
+					? ((stm32_ospi_hal_address_size(dev) ==
+					HAL_OSPI_ADDRESS_24_BITS)
+						? SPI_NOR_CMD_PP
+						: SPI_NOR_CMD_PP_4B)
+					: SPI_NOR_OCMD_PAGE_PRG;
```
[the remark this answers is not in the document]
Comment (FRASTM) [reply drivers/flash/flash_stm32_ospi.c:1048]: do you suggest
```
	if (dev_cfg->data_rate == OSPI_STR_TRANSFER) {
		s_command.Instruction = (dev_cfg->data_mode == OSPI_SPI_MODE)
					? ((stm32_ospi_hal_address_size(dev) ==
					HAL_OSPI_ADDRESS_24_BITS)
						? SPI_NOR_CMD_PP
						: SPI_NOR_CMD_PP_4B)
					: SPI_NOR_OCMD_PAGE_PRG;
	} else {
		s_command.Instruction = write_opcode;
```

File: drivers/flash/flash_stm32.c [hunk 7]
```diff
@@ -157,8 +157,7 @@ static void flash_stm32_flush_caches(const struct device *dev,
 		regs->ACR |= FLASH_ACR_DCEN;
 	}
 #elif defined(CONFIG_SOC_SERIES_STM32F7X)
-	SCB_InvalidateDCache_by_Addr((uint32_t *)(CONFIG_FLASH_BASE_ADDRESS
-						  + offset), len);
+	SCB_InvalidateDCache_by_Addr((uint32_t *)(FLASH_STM32_BASE_ADDRESS
 #endif
```
Comment (GeorgeCGV) [inline drivers/flash/flash_stm32.c:161]: missing `offset` and `len`?

Comment (FRASTM) [reply drivers/flash/flash_stm32.c:161]: yes, that will be rebased as soon as te https://github.com/zephyrproject-rtos/zephyr/pull/67989 is merged

File: drivers/flash/flash_stm32_ospi.c (hunk 5 above)
Comment (GeorgeCGV) [reply drivers/flash/flash_stm32_ospi.c]: Something like

```
	if (dev_cfg->data_rate == OSPI_DTR_TRANSFER) {
		/* DTR transfer rate (==> Octal mode) */
		cmd->Instruction = SPI_NOR_OCMD_DTR_RD;
		cmd->DummyCycles = SPI_NOR_DUMMY_RD_OCTAL_DTR;
	} else {
		/* STR transfer rate */
		if (dev_cfg->data_mode == OSPI_OPI_MODE) {
			/* OPI and STR */
			cmd->Instruction = SPI_NOR_OCMD_RD;
			cmd->DummyCycles = SPI_NOR_DUMMY_RD_OCTAL;
		} else {
			/* use SFDP:BFP read instruction */
			cmd->Instruction = dev_data->read_opcode;
			cmd->DummyCycles = dev_data->read_dummy_cycles;
		}
	}
```
At init stage the read opcode, dummy cycles, write opcode are supposed to be set. Some from SFDP:BFP and/or DT.

File: drivers/flash/flash_stm32_ospi.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (GeorgeCGV) [reply drivers/flash/flash_stm32_ospi.c:1048]:
```
	if (dev_cfg->data_rate == OSPI_STR_TRANSFER) {
		s_command.Instruction = (dev_cfg->data_mode == OSPI_SPI_MODE)
					? dev_data->write_opcode
					: SPI_NOR_OCMD_PAGE_PRG;
	} else {
		s_command.Instruction = SPI_NOR_OCMD_PAGE_PRG;
```

The `spi_nor_process_bfp` sets `write_opcode` accounting to mode if it is not set from DT. The address width is taken into consideration as well.

Also for write and read the lines might need adaptation not only based on mode, but also on the opcode. Something like:

**write**

```
if (dev_cfg->data_mode == OSPI_OPI_MODE) {
	cmd->AddressSize = HAL_OSPI_ADDRESS_32_BITS;
	cmd->InstructionSize = HAL_OSPI_INSTRUCTION_16_BITS;

	cmd->InstructionMode = HAL_OSPI_INSTRUCTION_8_LINES;
	cmd->AddressMode = HAL_OSPI_ADDRESS_8_LINES;
	cmd->DataMode = HAL_OSPI_DATA_8_LINES;
} else {
	cmd->AddressSize = stm32_ospi_hal_address_size(dev);
	cmd->InstructionSize = HAL_OSPI_INSTRUCTION_8_BITS;

	switch (cmd->Instruction) {
	case SPI_NOR_CMD_PP_4B:
		__fallthrough;
	case SPI_NOR_CMD_PP:
		cmd->InstructionMode = HAL_OSPI_INSTRUCTION_1_LINE;
		cmd->AddressMode = HAL_OSPI_ADDRESS_1_LINE;
		cmd->DataMode = HAL_OSPI_DATA_1_LINE;
		break;
	case SPI_NOR_CMD_PP_1_1_4_4B:
		__fallthrough;
	case SPI_NOR_CMD_PP_1_1_4:
		cmd->InstructionMode = HAL_OSPI_INSTRUCTION_1_LINE;
		cmd->AddressMode = HAL_OSPI_ADDRESS_1_LINE;
		cmd->DataMode = HAL_OSPI_DATA_4_LINES;
		break;
	case SPI_NOR_CMD_PP_1_4_4_4B:
		__fallthrough;
	case SPI_NOR_CMD_PP_1_4_4:
		cmd->InstructionMode = HAL_OSPI_INSTRUCTION_1_LINE;
		cmd->AddressMode = HAL_OSPI_ADDRESS_4_LINES;
		cmd->DataMode = HAL_OSPI_DATA_4_LINES;
		break;
	default:
		return -ENOTSUP;
	}
}
```

**read**

```
/* initially set lines by mode */
switch (dev_cfg->data_mode) {
case OSPI_QUAD_MODE:
	cmd->InstructionMode = HAL_OSPI_INSTRUCTION_4_LINES;
	cmd->AddressMode = HAL_OSPI_ADDRESS_4_LINES;
	cmd->DataMode = HAL_OSPI_DATA_4_LINES;
	break;
case OSPI_DUAL_MODE:
	cmd->InstructionMode = HAL_OSPI_INSTRUCTION_2_LINES;
	cmd->AddressMode = HAL_OSPI_ADDRESS_2_LINES;
	cmd->DataMode = HAL_OSPI_DATA_2_LINES;
	break;
default:
	return -ENOTSUP;
}

/* adapt by read mode */
switch (dev_data->read_mode) {
case JESD216_MODE_112:
	cmd->InstructionMode = HAL_OSPI_INSTRUCTION_1_LINE;
	cmd->AddressMode = HAL_OSPI_ADDRESS_1_LINE;
	cmd->DataMode = HAL_OSPI_DATA_2_LINES;
	break;
case JESD216_MODE_122:
	cmd->InstructionMode = HAL_OSPI_INSTRUCTION_1_LINE;
	cmd->AddressMode = HAL_OSPI_ADDRESS_2_LINES;
	cmd->DataMode = HAL_OSPI_DATA_2_LINES;
	break;
case JESD216_MODE_114:
	cmd->InstructionMode = HAL_OSPI_INSTRUCTION_1_LINE;
	cmd->AddressMode = HAL_OSPI_ADDRESS_1_LINE;
	cmd->DataMode = HAL_OSPI_DATA_4_LINES;
	break;
case JESD216_MODE_144:
	cmd->InstructionMode = HAL_OSPI_INSTRUCTION_1_LINE;
	cmd->AddressMode = HAL_OSPI_ADDRESS_4_LINES;
	cmd->DataMode = HAL_OSPI_DATA_4_LINES;
	break;
default:
	return -ENOTSUP;
}
```

Similar is done for normal operations. Therefore, some refactoring could help.

File: drivers/flash/flash_stm32_ospi.c [hunk 8]
```diff
@@ -1227,6 +1403,28 @@ static int flash_stm32_ospi_write(const struct device *dev, off_t addr,
...
 
+#ifdef CONFIG_STM32_MEMMAP
+	if (stm32_ospi_is_memorymap(dev)) {
+		/* If the Flash is in MemoryMapped mode, abort it and continue */
+		LOG_INF("MemoryMap : disable before write");
+		if (stm32_ospi_abort_memmap(dev) != 0) {
+			LOG_ERR("MemoryMap not aborted correctly");
+			return -EIO;
+		}
+	}
```
Comment (GeorgeCGV) [inline drivers/flash/flash_stm32_ospi.c:1414]: It would be fatal when accidently writing while running in XIP.

I guess the idea here is to allow bootloader to exit memory map mode that is entered in OSPI init to allow erase & write operations.

In that case, the `CONFIG_MCUBOOT` is going to be set to `y` for the mcuboot while the app won't have it. Extra macro could be used to prevent memory map mode exit in the application code.

An alternative, protect write abortion with extra Kconfig option that the application can set.

Comment (erwango) [reply drivers/flash/flash_stm32_ospi.c:1414]: Another alternative is to run flash driver from RAM. Not proposed today as not yet fully functional on our side, but this is an option used by others.

Comment (GeorgeCGV) [reply drivers/flash/flash_stm32_ospi.c:1414]: I am not sure what is the footprint on RAM in that case. But having flash driver in ram might not be required for all applications. After switching to the slot in o/qSPI it is rarely when writing and erasing are required outside of the bootloader.

**Comment by FRASTM** [comment]
see #68607
